core/cli/src/lib.rs

```diff
@@ -666,14 +679,10 @@ where
 		)?
 	}
 
-	let base_path = base_path(&cli.shared_params, version);
-
-	config.keystore_path = cli.keystore_path.unwrap_or_else(
-		|| keystore_path(&base_path, config.chain_spec.id())
-	);
+	config.keystore_path = cli.keystore_path.unwrap_or_else(|| config.in_chain_config_dir("keystore"));
 
 	config.database = DatabaseConfig::Path {
-		path: db_path(&base_path, config.chain_spec.id()),
+		path: config.database_path(),
```

Comment:
This line is kind of a red flag. We store `config.database_path()` in `config.database.path`.

Reply:
the general change of this PR is to move the generation of relative paths from the internal `*_path(&base_path`-style onto the configuration object. `database_path` there is only a handy alias for `self.base_path.clone().join("database")`. 

What exactly is the red flag?